scripts/elasticsearch_nodes.py

```diff
@@ -0,0 +1,130 @@
+current deployment that has two ES nodes:
+{"nodes": "[\"tVsGH2BSXXYe9sMLEYRcIv\", \"ydG9meZMAGkTnAlPgLljz8\", \"blank\", \"blank\"]"}
+"""
+from itertools import (
+    repeat,
+)
+import json
+import logging
+from time import (
+    sleep,
+)
+from typing import (
+    Sequence,
+)
+
+from botocore.exceptions import (
+    ClientError,
+)
+import jq
+
+from azul import (
+    config,
+    require,
+)
+from azul.deployment import (
+    aws,
+)
+from azul.logging import (
+    configure_script_logging,
+)
+
+log = logging.getLogger(__name__)
+
+# How many times to try getting the node IDs from AWS before giving up.
+max_attempts = 10
+
+
+def main() -> None:
+    node_ids = get_node_ids()
+    data = prepare_node_ids(node_ids)
+    print(json.dumps(data))
+
+
+def get_node_ids() -> Sequence[str]:
+    """
+    Return a list of ES node IDs used by the current deployment.
+    """
+    for i in range(max_attempts):
+        log.debug(f'Attempt {i}/{max_attempts}')
+        try:
+            response = aws.cloudwatch.list_metrics(Namespace='AWS/ES',
+                                                   MetricName='CPUUtilization',
+                                                   RecentlyActive='PT3H',
+                                                   Dimensions=[
+                                                       {
+                                                           'Name': 'DomainName',
+                                                           'Value': config.es_domain
+                                                       }
+                                                   ])
+        except ClientError as e:
+            log.warning(e.response['Error']['Code'] + '. Retrying...')
+        else:
+            nodes = [
+                d['Value']
+                for m in response['Metrics']
+                for d in m['Dimensions']
+                if d['Name'] == 'NodeId'
+            ]
+            if len(nodes) > 0:
+                return nodes
+            else:
+                log.warning('No node IDs found in response.')
+        sleep(5)
+
+    raise RuntimeError('Failed to obtain node IDs. Max attempts exhausted.')
+
+
+def prepare_node_ids(node_ids: Sequence[str]) -> dict[str, str]:
+    """
+    Return the given node IDs in a data structure that TerraForm can use as an
+    external data source.
+    """
+    # The list of nodes IDs is appended with dummy entries to ensure the length
+    # matches the number of placeholders in the template. If our list had less
+    # than the template, Terraform would throw an "Invalid index" error. Also,
+    # the dummy entries cannot be empty strings.
+    placeholder_count = node_placeholder_count()
+    require(placeholder_count > 0,
+            'Error: No node ID placeholders found in template file.')
+    node_ids = sorted(node_ids)
+    node_ids += list(repeat('blank', placeholder_count - len(node_ids)))
```

Comment:
Use https://more-itertools.readthedocs.io/en/latest/api.html#more_itertools.padded

Use `''` instead of `'blank'` if possible or briefly explain why it's not.

Reply:
Using an empty string value as padding fails during deploy:
```
aws_cloudwatch_dashboard.azul-dashboard-daniel: Modifying... [id=azul-dashboard-daniel]

Error: Putting dashboard failed: InvalidParameterInput: The dashboard body is invalid, there are 8 validation errors:
[
  {
    "dataPath": "/widgets/3/properties/metrics/7/5",
    "message": "Invalid metric field type, only \"String\" type is allowed"
  },
  ...
```
If the dashboard is modified manually via the AWS console an empty string value does work for some widgets, however those with metric expressions such as `ES shards` display an error message like `Error in expression e1 [ID 'm8' not found]`.

![1](https://user-images.githubusercontent.com/46407464/190597014-df7676dc-2f09-47cb-b464-fb48cdde5ab8.png)

Using only whitespace (' ') allows the deploy to complete, however the widgets then fail with the message `There was an error while trying to get graph data.`

![2](https://user-images.githubusercontent.com/46407464/190597070-09d0ba5c-2098-4ad4-9a98-daddf7c3c33c.png)

I will modify my comment in `prepare_node_ids()` to better explain the need for the "blank" value.